tests/test_convert_html.py

```diff
@@ -0,0 +1,271 @@
+    """
+
+    actual = html_to_htpy(input)
+    assert actual == """p["Special characters: & < > " ' ©"]"""
+
+
+def test_convert_script_style_tags():
+    input = """
+        <script type="text/javascript">alert('This is a script');</script>
+        <style>body { background-color: #fff; }</style>
+    """
+
+    actual = html_to_htpy(input, format=True)
+    assert actual == _format(
+        """[
+            script(type="text/javascript")[
+                "alert('This is a script');"
+            ],
+            style["body { background-color: #fff; }"],
+        ]"""
+    )
+
+
+def test_convert_html_doctype():
+    input = """
+        <!DOCTYPE html>
+        <html>
+        <head>
+          <title>Test Document</title>
+        </head>
+        <body>
+          <h1>Header</h1>
+          <p>Paragraph</p>
+        </body>
+        </html>
+    """
+
+    actual = html_to_htpy(input)
+    expected = (
+        """html[head[title["Test Document"]],body[h1["Header"],p["Paragraph"]]]"""
+    )
+
+    assert actual == expected
+
+
+def test_convert_empty_elements():
+    input = """
+        <div></div>
+        <p></p>
+        <span></span>
+    """
+
+    actual = html_to_htpy(input)
+    assert actual == "[div,p,span]"
+
+
+def test_convert_custom_tag():
+    input = """
+        <custom-element attribute="value">Custom content</custom-element>
+    """
+
+    actual = html_to_htpy(input)
+    assert actual == """custom_element(attribute="value")["Custom content"]"""
+
+
+def test_convert_malformed_html():
+    input = """
+        <div>
+          <p>Paragraph without closing tag
+          <div>Another div</p>
+        </div>
+    """
+
+    with pytest.raises(Exception) as e:
+        html_to_htpy(input)
+
+    assert "Closing tag p does not match the currently open tag (div)" in str(e.value)
+
+
+def test_convert_attributes_without_values():
+    input = """
+        <input type="checkbox" checked />
+        <option selected>Option</option>
+    """
+    actual = html_to_htpy(input)
+    assert (
+        actual
+        == """[input(type="checkbox",checked=True),option(selected=True)["Option"]]"""
+    )
+
+
+def test_convert_section_regular():
+    input = """
+        <section class="hero is-fullheight is-link">
+          <div class="hero-body">
+            <div class='container'>
+              <p class="subtitle is-3 is-spaced">Welcome</p>
+              <p class="title is-1 is-spaced">Student code: {{student_code}}</p>
+            </div>
+          </div>
+        </section>
+    """
+
+    actual = html_to_htpy(input, shorthand_id_class=False, format=True)
+    expected = _format(
+        """
+        section(class_="hero is-fullheight is-link")[
+            div(class_="hero-body")[
+                div(class_="container")[
+                    p(class_="subtitle is-3 is-spaced")["Welcome"],
+                    p(class_="title is-1 is-spaced")[f"Student code: {student_code}"]
```

Comment:
While flashy, it comes with a couple of challanges, I realize after the fact: 
1. a need to escape certain symbols that may exist in the content, e.g `"`, `{` and `}`. There might also be more I'm forgetting (?). Ill add these in a commit shortly.
2. Currently this is only done to content, not for example to attributes. Do you think that is sufficient, or would you prefer doing this to props as well?

Reply:
Maybe it would be better to try to use `'` or `"` (whichever is not used in the content) if one is 'available', and only escape in the last case, where both of them are already in use. But I don't know, don't feel like I personally use double quotes in html content that often so maybe its not terrible that they are escaped by the tool. Regarding the `"` symbol this obviously also applies to all content, not just template expressions.

Reply:
I believe it's covered now by this test:

```python
def test_convert_f_string_escaping():
    input = """
        <p>{{ variable }} is "a" { paragraph }.</p>
    """

    actual = html2htpy(input, format=False)
    expected = r'p[f"{ variable } is \"a\" {{ paragraph }}."]'

    assert actual == expected
```

Reply:
I think it is fine even though it may not be perfect. We can always improve it in the future too. It is not crucial to get all edge cases right, the generated code would still need to be adopted by hand to be fully usable.